cmd/zc_traverser_file.go

```diff
@@ -142,47 +147,64 @@ func (t *fileTraverser) traverse(preprocessor objectMorpher, processor objectPro
 	// So include the root dir/share in the enumeration results, if it exists or is just the share root.
 	_, err = directoryURL.GetProperties(t.ctx)
 	if err == nil || targetURLParts.DirectoryOrFilePath == "" {
-		err = processEntity(newAzFileRootFolderEntity(&directoryURL, ""))
+		s, err := convertToStoredObject(newAzFileRootFolderEntity(directoryURL, ""))
+		if err != nil {
+			return err
+		}
+		err = processStoredObject(s.(storedObject))
 		if err != nil {
 			return err
 		}
 	}
 
-	// Enumerate its contents
-	dirStack := &directoryStack{}
-	dirStack.Push(directoryURL)
-	for currentDirURL, ok := dirStack.Pop(); ok; currentDirURL, ok = dirStack.Pop() {
-		// Perform list files and directories.
+	// Define how to enumerate its contents
+	// This func must be threadsafe/goroutine safe
+	enumerateOneDir := func(dir parallel.Directory, enqueueDir func(parallel.Directory), enqueueOutput func(parallel.DirectoryEntry)) error {
+		currentDirURL := dir.(azfile.DirectoryURL)
 		for marker := (azfile.Marker{}); marker.NotDone(); {
 			lResp, err := currentDirURL.ListFilesAndDirectoriesSegment(t.ctx, marker, azfile.ListFilesAndDirectoriesOptions{})
 			if err != nil {
 				return fmt.Errorf("cannot list files due to reason %s", err)
 			}
-
-			// Process the files and folders we listed
-			fs := make([]azfileEntity, 0, len(lResp.FileItems)+len(lResp.DirectoryItems))
 			for _, fileInfo := range lResp.FileItems {
-				fs = append(fs, newAzFileFileEntity(currentDirURL, fileInfo))
+				enqueueOutput(newAzFileFileEntity(currentDirURL, fileInfo))
 			}
 			for _, dirInfo := range lResp.DirectoryItems {
-				fs = append(fs, newAzFileChildFolderEntity(currentDirURL, dirInfo.Name))
-			}
-			for _, f := range fs {
-				processErr := processEntity(f)
-				if processErr != nil {
-					return processErr
-				}
-			}
-
-			// If recursive is turned on, add sub directories.
-			if t.recursive {
-				for _, dirInfo := range lResp.DirectoryItems {
-					d := currentDirURL.NewDirectoryURL(dirInfo.Name)
-					dirStack.Push(d)
+				enqueueOutput(newAzFileChildFolderEntity(currentDirURL, dirInfo.Name))
+				if t.recursive {
+					// If recursive is turned on, add sub directories to be processed
+					enqueueDir(currentDirURL.NewDirectoryURL(dirInfo.Name))
 				}
 			}
 			marker = lResp.NextMarker
 		}
+		return nil
+	}
+
+	// run the actual enumeration.
+	// First part is a parallel directory crawl
+	// Second part is parallel conversion of the directories and files to stored objects. This is necessary because the conversion to stored object may hit the network and therefore be slow in not parallelized
+	parallelism := 1
+	if enumerationParallelism > 1 {
+		parallelism = enumerationParallelism / 2 // half for crawl, half for transform
+	}
```

Comment:
I wonder if it'd be preferable to do 1/3 or 1/4th for transform. My guess is that network calls are more costly than the actual transform.

Reply:
I have not yet managed to figure out how much of the network activity is in the transform (where we make the get properties call) and how much in the directory enumeration - I suspect the balance will vary based on directory structure.  So will leave at 1/2 for now.